ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java

```diff
@@ -819,23 +819,25 @@ private CompletableFuture<RaftClientReply> appendTransaction(
 
     assertLifeCycleState(LifeCycle.States.RUNNING);
 
+    final LeaderStateImpl leaderState = role.getLeaderStateNonNull();
+    final PendingRequests.Permit permit = leaderState.tryAcquirePendingRequest(request.getMessage());
```

Comment:
Let's call it `unsyncedPermit` since, if the leader state has changed, we need to get the permit again.
```java
    final PendingRequests.Permit unsyncedPermit = unsyncedLeaderState.tryAcquirePendingRequest(request.getMessage());
    if (unsyncedPermit == null) {
      return getResourceUnavailableReply(request, cacheEntry);
    }
```

```java
  private CompletableFuture<RaftClientReply> getResourceUnavailableReply(RaftClientRequest request, CacheEntry entry) {
    return entry.failWithException(new ResourceUnavailableException(
        getMemberId() + ": Failed to acquire a pending write request for " + request));
  }
```